### ISSUE TITLE
IAMPolicy for StorageBucket > cannot set projectEditor / projectOwner / projectViewer in spec.bindings.members

### ISSUE THREAD

**Issue by jlgeering** [GoogleCloudPlatform/k8s-config-connector#208]
**Describe the bug**
We want to create an iam.cnrm.cloud.google.com/IAMPolicy for a storage.cnrm.cloud.google.com/StorageBucket but the policy gets rejected because the members value does not match the validation regex

**ConfigConnector Version**
1.72

**To Reproduce**
*Steps to reproduce the behavior:*

* create a new bucket (see below), kubectl apply
* download the default policy with gsutil
* create an equivalent policy (see below), kubectl apply

*YAML snippets:*

```yaml
apiVersion: storage.cnrm.cloud.google.com/v1beta1
kind: StorageBucket
metadata:
  name: unique-bucket-name
```

output of: `gsutil iam get gs://unique-bucket-name`

```json
{
  "bindings": [
    {
      "members": [
        "projectEditor:project-name-123",
        "projectOwner:project-name-123"
      ],
      "role": "roles/storage.legacyBucketOwner"
    },
    {
      "members": [
        "projectViewer:project-name-123"
      ],
      "role": "roles/storage.legacyBucketReader"
    }
  ],
  "etag": "CBo="
}
```

```yaml
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPolicy
metadata:
  name: policy-for-bucket
spec:
  resourceRef:
    apiVersion: storage.cnrm.cloud.google.com/v1beta1
    kind: StorageBucket
    name: unique-bucket-name
  bindings:
    - members:
        - projectEditor:project-name-123
        - projectOwner:project-name-123
      role: roles/storage.legacyBucketOwner
    - members:
        - projectViewer:project-name-123
      role: roles/storage.legacyBucketReader
```

error message:

`
IAMPolicy.iam.cnrm.cloud.google.com "policy-for-bucket" is invalid: spec.bindings.members: Invalid value: "": spec.bindings.members in body should match '^(user|serviceAccount|group|domain):.+|allUsers|allAuthenticatedUsers$'
`

**Comment by spew** [comment]
Hello @jlgeering, `projectEditor:project-name-123` does not appear to be a valid GCP identity. The Members field must be filled with a value that starts with one of the following:
- user:
- serviceAccount:
- group:
- domain:

You can read more here: https://cloud.google.com/iam/docs/reference/rest/v1/Policy#Binding

**Comment by jlgeering** [comment]
hi @spew 

that is exactly the problem: storage.googleapis.com/Bucket assets use "invalid" values!

another example: create a bucket using the console, export using cloud asset api.

```json
{
    "name": "//storage.googleapis.com/__EDITED__",
    "asset_type": "storage.googleapis.com/Bucket",
    "iam_policy": {
        "etag": "BwWjZMFjHWE=",
        "bindings": [
            {
                "role": "roles/storage.legacyBucketOwner",
                "members": [
                    "projectEditor:__EDITED__",
                    "projectOwner:__EDITED__"
                ]
            },
            {
                "role": "roles/storage.legacyBucketReader",
                "members": [
                    "projectViewer:__EDITED__"
                ]
            }
        ]
    },
    "ancestors": __EDITED__,
    "update_time": "2020-04-16T09:07:00.108641Z"
}
```

**Comment by jlgeering** [comment]
to explain a little bit better: the iam_policy you see above was not created by me ... this seems to be the default policy when creating a bucket

once we overwrite this policy, there seems to be no way of getting back to the default policy, because it contains "invalid" values for the members field

**Comment by spew** [comment]
Got it, that makes sense. Apologize for not understanding the bucket default policy. We will look more into this issue and let you know when we have an update.

Is this issue blocking you or can you work around it for now?

**Comment by jlgeering** [comment]
we can work around it

**Comment by AlexBulankou** [comment]
thanks, @jlgeering . Another question: after your create bucket using Config Connector:
```yaml
apiVersion: storage.cnrm.cloud.google.com/v1beta1
kind: StorageBucket
metadata:
  name: unique-bucket-name
```

what does the output show when you run `kubectl describe storagebucket unique-bucket-name` ?

Trying to understand, 
if the bucket created successfully initially with Config Connector and then you are having problems trying to update the settings? OR
you start getting errors when reconciliation loop starts to run, without even updating any settings?
Thanks!

**Comment by jlgeering** [comment]
There are no problems whatsoever with the bucket itself, or with the management of the bucket through config connector. The problem is the the iam policy for the bucket which contains "invalid" values. And I am putting "invalid" in quotes, because I do not think that the values are invalid, but rather that the parameter validation for the members field is incomplete.

(I'm guessing that it might be a problem of the underlying rest api, and not directly a config connector issue, based on this link https://cloud.google.com/iam/docs/reference/rest/v1/Policy#Binding)

```
➜ kubectl describe storagebucket e29b3ea9-7669-4929-8ee0-bef492f5db23

Name:         e29b3ea9-7669-4929-8ee0-bef492f5db23
Namespace:    cc-ufg-play
Annotations:  cnrm.cloud.google.com/management-conflict-prevention-policy: resource
              cnrm.cloud.google.com/project-id: silver-pad-223012
              kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"storage.cnrm.cloud.google.com/v1beta1","kind":"StorageBucket","metadata":{"annotations":{},...
API Version:  storage.cnrm.cloud.google.com/v1beta1
Kind:         StorageBucket
Metadata:
  Creation Timestamp:  2020-06-16T20:28:24Z
  Finalizers:
    cnrm.cloud.google.com/finalizer
    cnrm.cloud.google.com/deletion-defender
  Generation:        3
  Resource Version:  9380070
  Self Link:         /apis/storage.cnrm.cloud.google.com/v1beta1/namespaces/cc-ufg-play/storagebuckets/e29b3ea9-7669-4929-8ee0-bef492f5db23
  UID:               b62a87e7-8ae9-4d08-a3e8-e2ec9abcb73d
Spec:
  Location:       US
  Storage Class:  STANDARD
Status:
  Conditions:
    Last Transition Time:  2020-06-16T20:28:25Z
    Message:               The resource is up to date
    Reason:                UpToDate
    Status:                True
    Type:                  Ready
  Self Link:               https://www.googleapis.com/storage/v1/b/e29b3ea9-7669-4929-8ee0-bef492f5db23
  URL:                     gs://e29b3ea9-7669-4929-8ee0-bef492f5db23
Events:
  Type    Reason    Age   From                      Message
  ----    ------    ----  ----                      -------
  Normal  Updating  36s   storagebucket-controller  Update in progress
  Normal  UpToDate  35s   storagebucket-controller  The resource is up to date

➜ gsutil iam get gs://e29b3ea9-7669-4929-8ee0-bef492f5db23
{
  "bindings": [
    {
      "members": [
        "projectEditor:silver-pad-223012",
        "projectOwner:silver-pad-223012"
      ],
      "role": "roles/storage.legacyBucketOwner"
    },
    {
      "members": [
        "projectViewer:silver-pad-223012"
      ],
      "role": "roles/storage.legacyBucketReader"
    }
  ],
  "etag": "CAE="
}
```

**Comment by jlgeering** [comment]
another way to look at it, is that the config connector cli generates invalid yaml for this bucket (see https://cloud.google.com/config-connector/docs/how-to/importing-existing-resources)

```
➜ echo '{"name":"//storage.googleapis.com/e29b3ea9-7669-4929-8ee0-bef492f5db23","asset_type":"storage.googleapis.com/Bucket","ancestors":["projects/652145480989"]}' | config-connector

---
apiVersion: storage.cnrm.cloud.google.com/v1beta1
kind: StorageBucket
metadata:
  annotations:
    cnrm.cloud.google.com/force-destroy: "false"
    cnrm.cloud.google.com/project-id: projects/652145480989
  labels:
    cnrm-lease-expiration: "1592341704"
    cnrm-lease-holder-id: brjm8q5fj76gu5msgef0
    managed-by-cnrm: "true"
  name: e29b3ea9-7669-4929-8ee0-bef492f5db23
spec:
  location: US
  storageClass: STANDARD
status:
  selfLink: https://www.googleapis.com/storage/v1/b/e29b3ea9-7669-4929-8ee0-bef492f5db23
  url: gs://e29b3ea9-7669-4929-8ee0-bef492f5db23
---
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPolicy
metadata:
  name: e29b3ea9-7669-4929-8ee0-bef492f5db23-iampolicy
spec:
  bindings:
  - members:
    - projectEditor:silver-pad-223012
    - projectOwner:silver-pad-223012
    role: roles/storage.legacyBucketOwner
  - members:
    - projectViewer:silver-pad-223012
    role: roles/storage.legacyBucketReader
  resourceRef:
    apiVersion: storage.cnrm.cloud.google.com/v1beta1
    kind: StorageBucket
    name: e29b3ea9-7669-4929-8ee0-bef492f5db23
status: {}
...
```

**Comment by spew** [comment]
Thanks for all the information @jlgeering -- the problem is our regexes that validate the values in 'members' are too restrictive for these values that the Storage service inserts.

**Comment by AlexBulankou** [comment]
As a temporary workaround you can edit iampolicymember CRD and remove [this pattern].(https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/crds/iam_v1beta1_iampolicymember.yaml#L65)

**Comment by spew** [comment]
This issue is fixed in the 1.13.0 release.

**Comment by jlgeering** [comment]
@spew it's better now, but I don't think it's 100% fixed yet:

`IAMPolicy.iam.cnrm.cloud.google.com "ufg-favur-assets" is invalid: spec.bindings.members: Invalid value: "": spec.bindings.members in body should match '^(user|serviceAccount|group|domain|projectEditor|projectOwner):.+|allUsers|allAuthenticatedUsers$'`

=> `projectViewer` is missing

**Comment by spew** [comment]
Hi @jlgeering thanks for reporting that. Another customer reported that and we fixed it in release 1.15.0:

https://github.com/GoogleCloudPlatform/k8s-config-connector/releases/tag/1.15.0

**Comment by jlgeering** [comment]
okidoki, thx
